v12 Migration Guide (#4862)
cypress-io/cypress-documentation: content/guides/references/migration-guide.md

title: Migration Guide
---

## Migrating to Cypress version 12.0

This guide details the changes and how to change your code to migrate to Cypress

> version 11.0.

Reviewer: ```suggestion
version 12.0.
```

> [See the full changelog for version 11.0](/guides/references/changelog#11-0-0).

Reviewer: ```suggestion
[See the full changelog for version 12.0](/guides/references/changelog#12-0-0).
```

title: Migration Guide
---

> ## Migrating to Cypress version 12.0

Reviewer 2: I'd like to add in here somewhere a section about retryability, linking to the rewritten guide.

> Cypress 12 has improved and changed the way we retry commands by introducing a new type of Cypress command: Queries. Many commands you're already familiar with are now queries, such as `.get()`, `.find()`, and `.children()`. You can learn more about queries - and the way they protect you from Detached From DOM errors - in our guide on [Retry-ability](https://deploy-preview-4835--cypress-docs.netlify.app/guides/core-concepts/retry-ability#Commands-Queries-and-Assertions).

> In most cases, you won't need to change your tests to take advantage of this functionality, but it does have one implication to be aware of, in the form of a new Best Practice for writing Cypress tests:

> - [Action commands](https://deploy-preview-4835--cypress-docs.netlify.app/guides/core-concepts/interacting-with-elements#Actionability) (such as `.click()`, `.type()`, and `.trigger()`) should be at the end of chains, not the middle.

> See the guide linked above for more details.

test and verify it can run successfully without the test before it.

### Alias Behaviors Changes

Cypress always re-queries aliases when they are referenced. This can result in
certain tests that use to pass failing. For example,

```js
cy.findByTestId('popover')
  .findByRole('button', { expanded: true })
  .as('button')
  .click()

cy.get('@button').should('have.attr', 'aria-expanded', 'false')
```

previously passed, because the initial button was collapsed when first queried,
and then later expanded. However, in Cypress 12, this test fails because the
alias is always re-queried from the DOM, effectively resulting in the following
execution:

```js
cy.findByTestId('popover').findByRole('button', { expanded: true }).click()

cy.findByTestId('popover')
  .findByRole('button', { expanded: true }) // A button which matches here (is expanded)...
  .should('have.attr', 'aria-expanded', 'false') // ...will never pass this assertion.
```

You can rewrite tests like this to be more specific; in our case, we changed the
alias to be the first button rather than the unexpanded button.

```js
cy.findByTestId('popover').findAllByRole('button').first().as('button')
```

> ### Command / Cypress API Changes

Reviewer 2: It feels like this should be at a different header level than the elements below it, or removed entirely (the ### headers below seem like they stand on their own just fine).

`.then()` instead of `.invoke()`.

```diff
cy.wrap(myAPI)
-  .invoke('makeARequest', 'http://example.com')
+  .then(api => api.makeARequest('http://example.com'))
   .then(res => { ...handle response... })
```

If `.invoke()` is followed by additional commands or assertions, it will call
the named function multiple times. This has the benefit that the chained
assertions can more reliably use the function's return value.

If this behavior is undesirable because you expect the function to invoked only
once, break the command chain and move the chains commands and/or assertions to
their own chain. For example, rewrite

```diff
- cy.get('input').invoke('val', 'text').type('newText')
+ cy.get('input').invoke('val', 'text')
+ cy.get('input').type('newText')
```

### `.should()`

`.should()` now throws an error if Cypress commands are invoked from inside a
`.should()` callback. This previously resulted in unusual and undefined
behavior. If you wish to execute series of commands on the yield value,
use`.then()` instead.

```diff
cy.get('button')
-  .should(($button) => {

    })
+  .then(api => api.makeARequest('http://example.com'))
   .then(res => { ...handle response... })
```

> If `.invoke()` is followed by additional commands or assertions, it will call

Reviewer 2: This section looks duplicated from ### invoke() above.

the named function multiple times. This has the benefit that the chained
assertions can more reliably use the function's return value.

If this behavior is undesirable because you expect the function to invoked only
once, break the command chain and move the chains commands and/or assertions to
their own chain. For example, rewrite

```diff
- .invoke('val', 'text').type('newText')
+ cy.get('input').invoke('val', 'text')
+ cy.get('input').type('newText')
```

#### `Cypress.Commands.override()`

> The follow commands can no longer be overridden:

Reviewer 2: ```suggestion
The follow commands can no longer be overridden with `Cypress.Commands.overwrite()`:
```

- .as()
- .children()
- .closest()
- .contains()
- cy.debug()
- cy.document()
- .eq()
- .filter()
- .find()
- .first()
- .focused()
- .get()
- .hash()
- .its()
- .last()
- cy.location()
- .next()
- .nextAll()
- .not()
- .parent()
- .parents()
- .parentsUntil()
- .prev()
- .prevUntil()
- cy.root()
- .shadow()
- .siblings()
- cy.title()
- cy.url()
- cy.window()

> 

Reviewer 2: ```suggestion

If you were previously ovrewriting them, instead add your version as a new command using `Cypress.Commands.add()` under a different name.

```

when using this mode and ensure tests continue to run independent from one
other.

<Badge type="danger">For example</Badge>the following tests are not independent
nor deterministic:

```js
describe('workflow', { testIsolation: 'off' }, () => {
  it('logs in', () => {
    cy.visit('my-app.com/log-in)
    cy.get('username').type('User1')
    cy.get('password').type(Cypress.env('User1_password'))
    cy.get('button#login').click()
    cy.contains('User1')
  })

  it('clicks user profile', () => {
    cy.get('User1').find('#profile_avatar).click()
    cy.contains('Email Preferences')
  })

  it('updates profile', () => {
    cy.get('button#edit')
    cy.get('email').type('user1@email.com')
    cy.get('button#save').click()
  })
})
```

In the above example, each test is relying on the previous test to be
_successful_ to correctly execute. If at any point, the first or second test
fails, the sequential test(s) will automatically fail and provided un-reliable
debugging errors since the errors are representative of the previous test.

The best way to ensure your tests are independent is to add a `.only()` to your
test and verify it can run successfully without the test before it.

> ### Alias Behaviors Changes

Reviewer 3: To me this is a slightly confusing heading. Maybe change to "Changes to behavior of aliases"?

+   cy.intercept('GET', /api/, (req) => {
+      req.on('response', (res) => {
+        res.setDelay(500)
+      })
+      req.body.'test': 'We’ll'
+    }).as('getApi')
    cy.visit('/index.html')
    cy.window().then((win) => {
      const xhr = new win.XMLHttpRequest
      xhr.open('GET', '/api/v1/foo/bar?a=42')
      xhr.send()
    })

    cy.wait('@getApi')
-   .its('url').should('include', 'api/v1')
+   .its('request.url').should('include', 'api/v1')
  })
```

#### `.invoke()`

[`.invoke()`](/api/commands/invoke) now throws an error if the function returns
a promise. If you wish to call a method that returns a promise and wait for it
to resolve, use [`.then()`](/api/commands/then) instead of `.invoke()`.

```diff
cy.wrap(myAPI)
-  .invoke('makeARequest', 'http://example.com')
+  .then(api => api.makeARequest('http://example.com'))
   .then(res => { ...handle response... })
```

If `.invoke()` is followed by additional commands or assertions, it will call
the named function multiple times. This has the benefit that the chained
assertions can more reliably use the function's return value.

If this behavior is undesirable because you expect the function to invoked only

> once, break the command chain and move the chains commands and/or assertions to

Reviewer 2: ```suggestion
once, break the command chain and move the chained commands and/or assertions to
```

> once, break the command chain and move the chains commands and/or assertions to

Reviewer 2: We want `chained` here rather than `chain's`. `chains` is definitely incorrect though. :D

> once, break the command chain and move the chain's commands and/or assertions to

Author: ```suggestion
once, break the command chain and move the chain'ed commands and/or assertions to
```

Reviewer 2: Don't want the apostrophe, chained not chain'ed

Author: 🤦🏻‍♀️ whoops. nice catch.

> once, break the command chain and move the chain'ed commands and/or assertions to

Author: ```suggestion
once, break the command chain and move the chained commands and/or assertions to
```

their own chain. For example, rewrite

```diff
- cy.get('input').invoke('val', 'text').type('newText')
+ cy.get('input').invoke('val', 'text')
+ cy.get('input').type('newText')
```

#### `.should()`

[`.should()`](/api/commands/should) now throws an error if Cypress commands are
invoked from inside a `.should()` callback. This previously resulted in unusual

> and undefined behavior. If you wish to execute series of commands on the yield

Reviewer 3: ```suggestion
and undefined behavior. If you wish to execute a series of commands on the yield
```

value, use`.then()` instead.

```diff
cy.get('button')
-  .should(($button) => {

    })
+  .then(api => api.makeARequest('http://example.com'))
   .then(res => { ...handle response... })
```

> #### `Cypress.Commands.override()`

Reviewer 2: ```suggestion
#### `Cypress.Commands.overwrite()`
```

Reviewer 2: Courtesy of one of our ambassadors who pointed this out, I would never have noticed myself. :)

Author: They should have commented! 😄 I'll update. Thanks for pointing this out.

In Cypress 12.0.0, we introduced a new command type, called queries. A query is
a small and fast command for getting data from the window or DOM. This
distinction is important because Cypress can retry chains of queries, keeping
the yielded subject up-to-date as a page rerenders.

With the introduction of query commands, the following commands have been

> re-categorized and can no longer be overridden with

Author: ```suggestion
re-categorized and can no longer be overwritten with
```

> [`Cypress.Commands.override()`](api/cypress-api/custom-commands#Overwrite-Existing-Commands):

Reviewer 2: ```suggestion
[`Cypress.Commands.overwrite()`](api/cypress-api/custom-commands#Overwrite-Existing-Commands):
```

- [`.as()`](/api/commands/as)
- [`.children()`](/api/commands/children)
- [`.closest()`](/api/commands/closest)
- [`.contains()`](/api/commands/contains)
- [`cy.debug()`](/api/commands/debug)
- [`cy.document()`](/api/commands/document)
- [`.eq()`](/api/commands/eq)
- [`.filter()`](/api/commands/filter)
- [`.find()`](/api/commands/find)
- [`.first()`](/api/commands/first)
- [`.focused()`](/api/commands/focused)
- [`.get()`](/api/commands/get)
- [`.hash()`](/api/commands/hash)
- [`.its()`](/api/commands/its)
- [`.last()`](/api/commands/last)
- [`cy.location()`](/api/commands/location)
- [`.next()`](/api/commands/next)
- [`.nextAll()`](/api/commands/nextall)
- [`.not()`](/api/commands/not)
- [`.parent()`](/api/commands/parent)
- [`.parents()`](/api/commands/parents)
- [`.parentsUntil()`](/api/commands/parentsuntil)
- [`.prev()`](/api/commands/prev)
- [`.prevUntil()`](/api/commands/prevuntil)
- [`cy.root()`](/api/commands/root)
- [`.shadow()`](/api/commands/shadow)
- [`.siblings()`](/api/commands/siblings)
- [`cy.title()`](/api/commands/title)
- [`cy.url()`](/api/commands/url)
- [`cy.window()`](/api/commands/window)

> If you were previously overwriting one the above commands, try adding your

Reviewer 3: ```suggestion
If you were previously overwriting one of the above commands, try adding your
```

    cy.get('#current-user').contains('logged in')
    cy.get('button#edit-1').click()
    cy.get('input#title').type('Updated title')
    cy.get('button#submit').click()
    cy.get('.toast').type('Changes saved!')
  })

  cy.visit('/items/1')
  cy.get('h1').contains('Updated title')
})
```

The just-released `cy.session()` command can be used to setup and cache cookies,
local storage and session storage between tests to easily re-establish the
previous (or common) browser contexts needed in a suite. This command will run
setup on its initial execution and will restore the saved browser state on each
sequential command execution. This command reduces the need for repeated
application logins, while users also benefit from the test isolation guardrails
to write independent, reliable and deterministic tests from the start.

If for whatever reason you still need to persist the dom and browser context
between tests, you can set `testIsolation='off` on the root configuration or at
the suite-level. For example:

```js
describe('workflow', { testIsolation: 'off' }, () => {
  ...
})
```

It is important to note that while turning test isolation `off` may improve the

> overall performance of end-to-end tests, however, it can also cause state to "leak"

Reviewer 4: ```suggestion
overall performance of end-to-end tests, it can however cause state to "leak"
```

that tests continue to run independently of one another.

<Badge type="danger">For example</Badge>the following tests are not independent
nor deterministic:

```js
describe('workflow', { testIsolation: 'off' }, () => {
  it('logs in', () => {
    cy.visit('my-app.com/log-in)
    cy.get('username').type('User1')
    cy.get('password').type(Cypress.env('User1_password'))
    cy.get('button#login').click()
    cy.contains('User1')
  })

  it('clicks user profile', () => {
    cy.get('User1').find('#profile_avatar).click()
    cy.contains('Email Preferences')
  })

  it('updates profile', () => {
    cy.get('button#edit')
    cy.get('email').type('user1@email.com')
    cy.get('button#save').click()
  })
})
```

In the above example, each test is relying on the previous test to be
_successful_ to correctly execute. If at any point, the first or second test
fails, the sequential test(s) will automatically fail and provide unreliable
debugging errors since the errors are representative of the previous test.

The best way to ensure your tests are independent is to add a `.only()` to your
test and verify it can run successfully without the test before it.

> ### Alias Behaviors Changes

Author: ```suggestion
### Behavior Changes in Alias Resolution
```

Cypress always re-queries aliases when they are referenced. This can result in

> certain tests that used to pass failing. For example,

Author: ```suggestion
certain tests that used to pass could start to fail. For example,
```

> certain tests that used to pass failing. For example,

Reviewer 4: I think you might mean "to start failing"
